apps/lib/apps_ui.c

```diff
@@ -99,18 +99,31 @@ static int ui_close(UI *ui)
     return 1;
 }
 
+static char *(ui_prompt_construct)(UI *ui, const char *object_desc,
+                                      const char *object_name)
+{
+    PW_CB_DATA *cb_data = (PW_CB_DATA *)UI_get0_user_data(ui);
+
+    return cb_data != NULL
+        ? UI_construct_prompt(NULL, "pass phrase", cb_data->prompt_info)
+        : UI_construct_prompt(NULL, object_desc, object_name);
+}
+
 int setup_ui_method(void)
 {
     ui_fallback_method = UI_null();
 #ifndef OPENSSL_NO_UI_CONSOLE
     ui_fallback_method = UI_OpenSSL();
 #endif
     ui_method = UI_create_method("OpenSSL application user interface");
-    UI_method_set_opener(ui_method, ui_open);
-    UI_method_set_reader(ui_method, ui_read);
-    UI_method_set_writer(ui_method, ui_write);
-    UI_method_set_closer(ui_method, ui_close);
-    return 0;
+    return ui_method != NULL
+        && 0 == UI_method_set_opener(ui_method, ui_open)
+        && 0 == UI_method_set_reader(ui_method, ui_read)
+        && 0 == UI_method_set_writer(ui_method, ui_write)
+        && 0 == UI_method_set_closer(ui_method, ui_close)
+        && 0 == UI_method_set_prompt_constructor(ui_method,
+                                                 ui_prompt_construct)
+        ? 0 : -1;
```

Comment:
Not your fault - but this has some messed up return codes :)

Reply:
> Not your fault - but this has some messed up return codes :)

Good that you pointed this out.
I've taken the liberty to change the return values of this (internal, undocumented) function 
to follow the typical pattern: 1 for success, 0 for failure.